### PR TITLE
refactor(storage): decouple compaction task picking and assignment.

### PR DESCRIPTION
**Pull request by zwang28**
## What's changed and what's your intention?
Previously we pick and assign a compaction task transactionally, 

- It may introduce unnecessary compaction task cancellation when a compactor is found unreachable later, considering picking a task is not that cheap. 
- It has also enforced a compactor's task queue' length to 1 (though can be fixed), which impedes us from making compactor multitasked.

This PR simplifies it:
1. Pick a compaction task.
2. Assign and send the task to a compactor transactionally. 
    - May try another compactor if previous one fails.
    - May block the compaction scheduler if no compactor available.
    - "send the task" can be time-consuming. It's fine as long as the held `RwLock<Compaction>` only guards compaction progress.
3. When meta failover, any pending compaction task which is not yet assigned is cancelled. (We could have avoided this step if we consume any pending task first in step 1, which requires the pending `CompactTask` is persistent. Because there can be at most on pending task, we prefer the occasional cancellation to frequent `CompactTask` writing)

TODO:

- Compactor should report progress so that stuck tasks can be cancelled.

### CODE REVIEW

File: src/meta/src/hummock/hummock_manager.rs [hunk 1]
```diff
@@ -541,42 +543,15 @@ where
...
-                .get(&versioning_guard.current_version_id.id())
-                .unwrap()
-                .clone()
-        };
+        let current_version = self.versioning.read().await.current_version();
         let compact_task = compact_status.get_compact_task(current_version.levels);
         let ret = match compact_task {
             None => Ok(None),
             Some(mut compact_task) => {
-                compact_task_assignment.insert(
```
Comment (zwang28) [inline src/meta/src/hummock/hummock_manager.rs:573]: This is a bug. Watermark should be set before insert. This PR has fixed it.

File: src/meta/src/hummock/compaction_scheduler.rs [hunk 2]
```diff
@@ -91,76 +93,79 @@ where
...
-                    .report_compact_task(&compact_task)
-                    .await
-                {
+        // 2. Assign the compaction task to a compactor.
+        'send_task: loop {
+            // 2.1 Select a compactor.
+            let compactor = match self.compactor_manager.next_compactor() {
+                None => {
+                    tracing::warn!("No compactor is available.");
+                    tokio::time::sleep(Duration::from_secs(60)).await;
```
Comment (Little-Wallace) [inline src/meta/src/hummock/compaction_scheduler.rs:121]: Will `60s` be too  long? In my experience, a compaction job shall complete in 1s ~ 30s.

Comment (Little-Wallace) [inline src/meta/src/hummock/compaction_scheduler.rs:121]: Will `60s` be too  long? In my experience, a compaction job shall complete in 1s ~ 30s.

**Review by Little-Wallace — APPROVED** [review]
rest LGTM

**Review by hzxa21 — APPROVED** [review]

**Comment by hzxa21** [comment]
> TODO:
> Compactor should report progress so that stuck tasks can be cancelled.

Another benifit of having regular progress report from compactor is that if a task is already assigned but not persisted, we just need to persist the assignment on receiving the report but not cancel the task.

File: src/meta/src/hummock/compaction_scheduler.rs (hunk 2 above)
Comment (zwang28) [reply src/meta/src/hummock/compaction_scheduler.rs:121]: This line means no compactor node is alive, before we support compactor node backpressure.
I think the interval you refer to will be introduced after backpressure.

**Pull request merged**